kml/serdes_gpx.cpp

```diff
@@ -358,6 +365,127 @@ std::string GpxParser::BuildDescription() const
   return m_description + "\n\n" + m_comment;
 }
 
+namespace
+{
+
+std::optional<std::string> GetDefaultLanguage(LocalizableString const & lstr)
+{
+  auto const firstLang = lstr.begin();
+  if (firstLang != lstr.end())
+    return {firstLang->second};
+  return {};
+}
+
+std::string CoordToString(double c)
+{
+  std::ostringstream ss;
+  ss.precision(8);
+  ss << c;
+  return ss.str();
+}
+
+void SaveColorToRGB(Writer & writer, uint32_t rgba)
+{
+  writer << NumToHex(static_cast<uint8_t>(rgba >> 24 & 0xFF))
+         << NumToHex(static_cast<uint8_t>((rgba >> 16) & 0xFF))
+         << NumToHex(static_cast<uint8_t>((rgba >> 8) & 0xFF));
+}
+
+
+void SaveCategoryData(Writer & writer, CategoryData const & categoryData)
+{
+  writer << "<metadata>\n";
+  auto name = GetDefaultLanguage(categoryData.m_name);
+  if (name.has_value())
+    writer << kIndent2 << "<name>" << name.value() << "</name>\n";
+  auto description = GetDefaultLanguage(categoryData.m_description);
+  if (description.has_value())
+  {
+    writer << kIndent2 << "<desc>";
+    SaveStringWithCDATA(writer, description.value());
+    writer << "</desc>\n";
+  }
+  writer << "</metadata>\n";
+}
+
+void SaveBookmarkData(Writer & writer, BookmarkData const & bookmarkData)
+{
+  auto const [lat, lon] = mercator::ToLatLon(bookmarkData.m_point);
+  writer << "<wpt lat=\"" << CoordToString(lat) << "\" lon=\"" << CoordToString(lon) << "\">\n";
+  auto name = GetDefaultLanguage(bookmarkData.m_name);
+  if (name.has_value())
+    writer << kIndent2 << "<name>" << name.value() << "</name>\n";
+  auto description = GetDefaultLanguage(bookmarkData.m_description);
+  if (description.has_value())
+  {
+    writer << kIndent2 << "<cmt>";
+    SaveStringWithCDATA(writer, description.value());
+    writer << "</cmt>\n";
+  }
+  writer << "</wpt>\n";
+}
+
+bool TrackHasAltitudes(TrackData const & trackData)
+{
+  auto const & lines = trackData.m_geometry.m_lines;
+  if (lines.empty())
+    return false;
+  auto const & line = lines.at(0);
+  if (line.empty())
+    return false;
+  return line.at(0).GetAltitude() != geometry::kDefaultAltitudeMeters
+         && line.at(0).GetAltitude() != geometry::kInvalidAltitude;
+}
+
+uint32_t TrackColor(TrackData const & trackData)
+{
+  if (trackData.m_layers.empty())
+    return kDefaultTrackColor;
+  return trackData.m_layers.at(0).m_color.m_rgba;
+}
+
+void SaveTrackData(Writer & writer, TrackData const & trackData)
+{
+  writer << "<trk>\n";
+  auto name = GetDefaultLanguage(trackData.m_name);
+  if (name.has_value())
+    writer << kIndent2 << "<name>" << name.value() << "</name>\n";
+  uint32_t color = TrackColor(trackData);
+  if (color != kDefaultTrackColor)
```

Comment:
C++17 allows to declare vars inside if:
```suggestion
  if (auto const color = TrackColor(trackData); color != kDefaultTrackColor)
```

Reply:
Fixed